From d36e72a4bf11a7e913966f0251cdfdc7d46a6707 Mon Sep 17 00:00:00 2001
From: Junping Du <junping_du@apache.org>
Date: Wed, 4 Feb 2015 19:10:52 -0800
Subject: [PATCH 1840/2748] HADOOP-10062. race condition in
 MetricsSystemImpl#publishMetricsNow that causes
 incorrect results.  (Contributed by Sangjin Lee)
 (cherry picked from commit
 0b567f424673b5cea1c3bc23b2bd268ef6b7625f)

(cherry picked from commit f59b698fc99211c010d569e7f71555183dfc29f1)

Change-Id: I08088a0657bafe1c30f513765d0aa2cf6200ab70
---
 .../hadoop/metrics2/impl/MetricsSystemImpl.java    |    2 +-
 .../metrics2/impl/TestMetricsSystemImpl.java       |    2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java
index 9752d30..422f9cd 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java
@@ -388,7 +388,7 @@ synchronized void onTimerEvent() {
    * Requests an immediate publish of all metrics from sources to sinks.
    */
   @Override
-  public void publishMetricsNow() {
+  public synchronized void publishMetricsNow() {
     if (sinks.size() > 0) {
       publishMetrics(sampleMetrics(), true);
     }    
diff --git a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/metrics2/impl/TestMetricsSystemImpl.java b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/metrics2/impl/TestMetricsSystemImpl.java
index b5ebb93..95cb9f0 100644
--- a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/metrics2/impl/TestMetricsSystemImpl.java
+++ b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/metrics2/impl/TestMetricsSystemImpl.java
@@ -192,7 +192,7 @@
       threads[i] = new Thread(new Runnable() {
         private boolean safeAwait(int mySource, CyclicBarrier barrier) {
           try {
-            barrier1.await(2, TimeUnit.SECONDS);
+            barrier.await(2, TimeUnit.SECONDS);
           } catch (InterruptedException e) {
             results[mySource] = "Interrupted";
             return false;
-- 
1.7.9.5

